### ISSUE TITLE
Question: Field.Schema v.s. Field.Validator = schema.Object

### ISSUE THREAD

**Issue by smyrman** [rs/rest-layer#77]
What are the differences between
```go
schema := schema.Schema{
        Fields: {
        }
}
```

**Comment by rs** [comment]
Good question. It was add by @yanfali In #24, he can certainly answer.

**Comment by yanfali** [comment]
Primarily use case. I wanted to use a schema in array types. Previously you could only use a type which didn't support schema.

**Comment by smyrman** [comment]
Ah, ok so `Field.Schema` is the initial support for sub-schemas here. I actually like the `Object` `FieldValidator` better than having `Field.Schema`, as it's both more consistent, and avoids ambiguous meaning like:

```go
 "meta": Field{Validator: schema.String{}, Schema: subSchema}
```

Would it be sensible to deprecate/remove the Schema attribute from Field - perhaps after tagging a release, as suggested in #65? I assume this could also simplify things internally and avoid handling the same case twice several places? For one I realize we don't handle Field.Schema in the `jsonschema` package.

**Comment by smyrman** [comment]
> For #65 I would rather wait for the Go official way to handle that.

OK, for the manifest part, but tagging a release in Git (e.g. v0.1.0), is something the team have already recommended us to do, but for end-users, rest-layer would be a lot safer/easier to use if you could rely on a tagged release. Especially when we do breaking changes.

**Comment by rs** [comment]
Sure we can tag.

**Comment by smyrman** [comment]
@rs - updated the title & description.

**Comment by rs** [comment]
Are you going to submit a PR for this one?

**Comment by smyrman** [comment]
I don't know if I can/want to figure out the implications on reference checking... there are some other easier issues I would rather look at, e.g. JSON schema ones...

**Comment by smyrman** [comment]
At least not know

**Comment by smyrman** [comment]
@rs Looking at the definition of Object:
```
type Object struct {
    Schema *Schema
}
```

Could we not make Schema implement FieldValidator or at least change the definition to:
```
type Object Schema
```

**Comment by smyrman** [comment]
Maybe it's not so easy btw. At least not without introducing _additional_ breaking changes everywhere...

**Comment by rs** [comment]
I'm all for introducing breaking changes now that simplify the API rather than later.

**Comment by smyrman** [comment]
I have done some background thinking on this, and I have a rather radical proposal:

- Move `schema.Fields` from `Schema` to `Object`, and let the Schema have a `Validator` field instead. We might want to somehow restrict the validator to support only Object types! However, we should at least be able to hold any of `schema.Dict`, `schema.Object`, `schema.AllOf`  and `schema.AnyOf` (where all entries are within the supported list, recursively checked, for the last two types). This may be hard-coded or based on interfaces.
- Remove `Schema` from `Field`.

Part of the reasoning for this, is that I have a use-case to support something like this at a top level:

```go
s := schema.Schema{
    Description: "My resource",
    Validator: &schema.AllOf{
        &schema.Object{
            Fields: commonFields,
        },
        $schema.AnyOf{
            $schema.Object{Fields: schema.Fields{
                "type":{Validator:&schema.String{Allowed:[]string{"x"}}}},
                "someField":{Validator: validatorX},
            },
            $schema.Object{Fields: schema.Fields{
                "type":{Validator:&schema.String{Allowed:[]string{"y"}}}},
                "someField":{Validator: validatorY},
            }
        }
    }
}

```
I can do this today via a `hook` with extra validator rools, but that would fail to generate the right documentation.

**Comment by smyrman** [comment]
The error type returned from `schema.Object` (and any other nested type), will be the existing `ErrorMap` type.

**Comment by rs** [comment]
Interesting idea. Then instead of `Validator` shouldn't we rename this field `Type`?

The root as Object might be enforced in `resource`. This restriction does not necessarily apply to all use-cases (outside of REST).

**Comment by smyrman** [comment]
IN #151, I am proposing to replace `ValuesValidator FieldValidator` in `schema.Dict`, `schema.Array`, `schema.AnyOf` and `schema.AllOf`. As already proposed in this issue, `schema.Object` will also contain a set of fields.

However, for all except `schema.Object`, keeping the `Required` property inside the `Field` definition becomes a bit awkward. JSON Schema have solved this ackwardness by making `required` a list of strings in the schema. Maybe we should consider to do the same for rest-layer. That would allow us to e.g. specify:

```go
Validator: $schema.Dict{
    Required: []string{"fieldA", "fieldB"},
    KeyValidator: &schema.String{},
    Values: schema.Field{...}
}
```

And also translate that to JSON Schema:
```
{
    "type": "object",
    "additionalProperties": {...},
    "required": ["fieldA", "fieldB"]
}
```

**Comment by smyrman** [comment]
> I have done some background thinking on this, and I have a rather radical proposal (...)

I want to drag this a _little bit_ further. I think the `schema.Field` and  `schema.Schema` types could be merged into one:

```go
type Object struct{
    Fields   map[string]Schema
    Required []string
}

type Object struct{
    Fields   map[string]Schema
    Required []string
}

type Schema {
    Title       string
    Description string
    ReadOnly    bool
    Validator   FieldValidatorBuilder
}
```
When it comes to implementation (some time in the future), align with ideas in #194.

**Comment by smyrman** [comment]
I plan to start this issue by:
- Creating a new (blank) repo for the schema package for experimental purposes.
- Copy over just enough types to get the design right.

I will probably take my time, and try to address as many design issues as possible, and then we can see what we want to do when it's done.

**Comment by rs** [comment]
Why not doing this in a branch?

**Comment by smyrman** [comment]
We use rest-layer actively. We are pretty satisfied with it, but over time we have hit most of the limitations regarding the current schema package at some point, and either worked around it, or left it in a. slightly undesirable state.

To get up with a desing that addresses all these. issues, I basically first want to proto-type it, without having to take the other rest-layer packages (or our own code) into account.

In this proto-type I want to attempt to solve this issue, and several other of issues that link to this one (e.g. #194 and #192, and maybe #138). Perhaps other issues that I can think of. The proto-type will not copy the query package. This may need to happen in several itteraitons.

Once the proto-type is done, I see two possible futures for the schema package:
1. We merge it back into rest-layer
2. We check if the package could possibly reach a larger user-base (and more potential contributers), if it's maintained as a separate package.

For case 2, the query package would necessarily remain in rest-layer, but except for that, there is not much dependencies on the schema validation/serialization package itself that can not be solved by a simple interface.

For the proto-type, weather we choose 1 or 2 is mostly irrelevant.

**Comment by smyrman** [comment]
Link to proto-type for new schema design:
- https://github.com/searis/schema

It's not in a working state.
